docs/site/Loopback-component-authorization.md

```diff
@@ -0,0 +1,550 @@
+user for any resource, which is known as **Authorization**. This is a
+minimalistic guide for creating such an implementation using Loopback component.
+This can be part of your main REST Application project or can be created as a
+Loopback extension for reuse in multiple projects. Latter is the better option
+for obvious reasons - reusability.
+
+**Note: Loopback team is working on making authorization an out-of-the-box
+feature in Loopback 4. It is a work in progress and will soon be there. Until
+then, this implementation guide can be followed.**
+
+## The requirement
+
+1. Every protected API end point needs to be restricted by specific permissions.
+2. API allows access only if logged in user has permission as per end point
+   restrictions.
+3. API throws **403 Forbidden** error if logged in user do not have sufficient
+   permissions.
+4. Publicly accessible APIs must be accessible regardless of user permissions.
+5. Every user has a set of permissions. These permissions may be associated via
+   role attached to the user or directly to the user.
+6. A user can be provided additional permissions or denied some permissions over
+   an above its role permissions. This is considered explicit allow/deny and
+   always takes precedence while calculating permissions.
+
+## Considerations
+
+There are a few considerations that are taken into account before this
+implementation can be done.
+
+1. User authentication is already implemented. You can refer to the
+   [@loopback/authentication](https://github.com/strongloop/loopback-next/tree/master/packages/authentication#loopbackauthentication)
+   guide.
+2. As part of authentication, client is sent back a token (JWT or similar) which
+   client need to pass in every API request headers thereafter.
+3. The authenticate action provider parses the token to return AuthResponse
+   object.
+4. AuthResponse contains the logged in user information including associated
+   role details.
+
+## The implementation
+
+First, let's define the types needed for this.
+
+{% include code-caption.html content="/src/authorization/types.ts" %}
+
+```ts
+import {PermissionKey} from './permission-key';
+
+/**
+ * Authorize action method interface
+ */
+export interface AuthorizeFn {
+  // userPermissions - Array of permission keys granted to the user
+  // This is actually a union of permissions picked up based on role
+  // attached to the user and allowed permissions at specific user level
+  (userPermissions: PermissionKey[]): Promise<boolean>;
+}
+
+/**
+ * Authorization metadata interface for the method decorator
+ */
+export interface AuthorizationMetadata {
+  // Array of permissions required at the method level.
+  // User need to have at least one of these to access the API method.
+  permissions: string[];
+}
+
+/**
+ * User Permission model
+ * used for explicit allow/deny any permission at user level
+ */
+export interface UserPermission {
+  permission: PermissionKey;
+  is_allowed: boolean;
```

Comment:
Let's use camel case, such as `isAllowed` or `allowed`.

Maybe it's better to use enum values, such as `allow | deny`.

Reply:
renamed to allowed. I would prefer boolean over enum as it avoid any possibility of 3rd value which I intend here.